src/ringbuffer/mod.rs

```diff
@@ -0,0 +1,241 @@
+use libc;
+use std;
+use jack_sys as j;
+
+pub struct RingBuffer(*mut j::jack_ringbuffer_t);
+
+impl RingBuffer {
+    pub fn new(size: usize) -> Result<Self, ()> {
+
+        let insize = size as libc::size_t;
+        let handle = unsafe { j::jack_ringbuffer_create(insize) };
+
+        if handle.is_null() {
+            return Err(());
+        }
+
+        Ok(RingBuffer(handle))
+    }
+
+    pub fn mlock(&mut self) {
+        unsafe { j::jack_ringbuffer_mlock(self.0) };
+    }
+
+    pub fn reset(&mut self) {
+        unsafe { j::jack_ringbuffer_reset(self.0) };
+    }
+
+    pub fn into_reader_writer(self) -> (RingBufferReader, RingBufferWriter) {
+        let handle = std::sync::Arc::new(self);
+
+        (RingBufferReader::new(handle.clone()), RingBufferWriter::new(handle))
+    }
+
+    pub fn from_reader_writer(r: RingBufferReader, w: RingBufferWriter) -> Self {
+        let mut handle_r = r.into_handle();
+        let handle_w = w.into_handle();
+
+        if handle_r.0 != handle_w.0 {
+            panic!("mismatching read and write handles!")
+        }
+
+        let handle = RingBuffer(handle_r.0);
+
+        // make sure that get_mut works
+        std::mem::drop(handle_w);
+        let muthandle = std::sync::Arc::get_mut(&mut handle_r).unwrap();
+        // make sure they don't close when they are dropped.
+        muthandle.0 = std::ptr::null_mut();
+        handle
+    }
+}
+
+
+impl Drop for RingBuffer {
+    fn drop(&mut self) {
+        if !self.0.is_null() {
+            unsafe { j::jack_ringbuffer_free(self.0) };
+        }
+        self.0 = 0 as *mut j::jack_ringbuffer_t;
+    }
+}
+
+pub struct RingBufferReader {
+    ringbuffer_handle: std::sync::Arc<RingBuffer>,
+}
+
+unsafe impl Send for RingBufferReader {}
+// impl !Sync for RingBufferReader{ }
+
+pub struct RingBufferWriter {
+    ringbuffer_handle: std::sync::Arc<RingBuffer>,
+}
+
+unsafe impl Send for RingBufferWriter {}
+// impl !Sync for RingBufferWriter{ }
+
+
+impl RingBufferReader {
+    fn new(handle: std::sync::Arc<RingBuffer>) -> Self {
```

Comment:
Would this be possible?

```rust
    let rb = Arc::new(RingBuffer::new(n).unwrap());
    let reader_a = RingBufferReader::new(rb.clone());
    let reader_b = RingBufferReader::new(rb.clone());
    let writer_a = RingBufferWriter::new(rb.clone());
    let writer_b = RingBufferWriter::new(rb.clone());
```
If so, the single reader, single writer precondition is broken.